### PR TITLE
added s3 keys generator

### PR DESCRIPTION
**Pull request by lizk886**
S3 key manager handles logics of generating S3 keys 
verified both s3encyrption_keys and metadata file can be successfully written to local stack

### CODE REVIEW

File: src/main/java/com/uid2/admin/Main.java [hunk 1]
```diff
@@ -119,7 +123,7 @@ public void run() {
             try {
                 adminKeysetProvider.loadContent();
             } catch (CloudStorageException e) {
-                if(e.getMessage().contains("The specified key does not exist")){
+                if (e.getMessage().contains("The specified key does not exist")) {
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/admin/Main.java:126]: Can you revert these?

File: src/main/java/com/uid2/admin/Main.java [hunk 2]
```diff
@@ -195,6 +199,15 @@ public void run() {
             operatorKeyProvider.loadContent(operatorKeyProvider.getMetadata());
             OperatorKeyStoreWriter operatorKeyStoreWriter = new OperatorKeyStoreWriter(operatorKeyProvider, fileManager, jsonWriter, versionGenerator);
 
+            CloudPath s3KeyMetadataPath = new CloudPath(config.getString(Const.Config.S3keysMetadataPathProp));
+            GlobalScope s3KeyGlobalScope = new GlobalScope(s3KeyMetadataPath);
+            RotatingS3KeyProvider s3KeyProvider = new RotatingS3KeyProvider(cloudStorage, s3KeyGlobalScope);
+            S3KeyStoreWriter s3KeyStoreWriter = new S3KeyStoreWriter(s3KeyProvider, fileManager, jsonWriter, versionGenerator, clock, s3KeyGlobalScope);
+            SecureKeyGenerator s3KeyGenerator = new SecureKeyGenerator();
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/admin/Main.java:206]: If this generator is just used by the manager, it should just declare it itself.

File: src/main/java/com/uid2/admin/Main.java [hunk 3]
```diff
@@ -195,6 +199,15 @@ public void run() {
             operatorKeyProvider.loadContent(operatorKeyProvider.getMetadata());
             OperatorKeyStoreWriter operatorKeyStoreWriter = new OperatorKeyStoreWriter(operatorKeyProvider, fileManager, jsonWriter, versionGenerator);
 
+            CloudPath s3KeyMetadataPath = new CloudPath(config.getString(Const.Config.S3keysMetadataPathProp));
+            GlobalScope s3KeyGlobalScope = new GlobalScope(s3KeyMetadataPath);
+            RotatingS3KeyProvider s3KeyProvider = new RotatingS3KeyProvider(cloudStorage, s3KeyGlobalScope);
+            S3KeyStoreWriter s3KeyStoreWriter = new S3KeyStoreWriter(s3KeyProvider, fileManager, jsonWriter, versionGenerator, clock, s3KeyGlobalScope);
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/admin/Main.java:205]: You will need to add a try catch like above to create the file

File: src/main/java/com/uid2/admin/managers/S3KeyManager.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    public S3KeyManager(RotatingS3KeyProvider s3KeyProvider, S3KeyStoreWriter s3KeyStoreWriter, IKeyGenerator keyGenerator) {
+        this.s3KeyProvider = s3KeyProvider;
+        this.s3KeyStoreWriter = s3KeyStoreWriter;
+        this.keyGenerator = keyGenerator;
+
+
+    }
+
+    public S3Key generateS3Key(int siteId, long activates, long created) throws Exception {
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/admin/managers/S3KeyManager.java:31]: Does this need to be public?

File: src/main/java/com/uid2/admin/managers/S3KeyManager.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            s3Keys = new HashMap<>();
+        } else {
+            s3Keys = new HashMap<>(s3Keys);
+        }
+        s3Keys.put(s3Key.getId(), s3Key);
+        s3KeyStoreWriter.upload(s3Keys, null);
+    }
+
+
+    protected int getNextKeyId() {
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/admin/managers/S3KeyManager.java:65]: This should be private

File: src/main/java/com/uid2/admin/managers/S3KeyManager.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        return keyGenerator.generateRandomKeyString(32);
+    }
+
+    public void addS3Key(S3Key s3Key) throws Exception {
+        Map<Integer, S3Key> s3Keys = new HashMap<>(s3KeyProvider.getAll());
+        s3Keys.put(s3Key.getId(), s3Key);
+        s3KeyStoreWriter.upload(s3Keys, null);
+    }
+
+    public void addOrUpdateS3Key(S3Key s3Key) throws Exception {
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/admin/managers/S3KeyManager.java:53]: I don't see addS3Key used. Also I don't think need to be public either

File: src/main/java/com/uid2/admin/managers/S3KeyManager.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        for (OperatorKey operatorKey : operatorKeys) {
+            uniqueSiteIds.add(operatorKey.getSiteId());
+        }
+
+        for (Integer siteId : uniqueSiteIds) {
+            if (!doesSiteHaveKeys(siteId)) {
+                for (int i = 0; i < keyCountPerSite; i++) {
+                    long created = Instant.now().getEpochSecond();
+                    long activated = created + (i * keyActivateInterval);
+                    addOrUpdateS3Key(generateS3Key(siteId, activated, created));
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/admin/managers/S3KeyManager.java:101]: I think we should be using just addKey here right? This should never be changing a key

File: src/main/java/com/uid2/admin/vertx/service/OperatorKeyService.java [hunk 8]
```diff
@@ -13,6 +13,7 @@
 import com.uid2.shared.auth.*;
 import com.uid2.shared.secret.KeyHashResult;
 import com.uid2.shared.secret.KeyHasher;
+import com.uid2.shared.store.reader.RotatingS3KeyProvider;
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/admin/vertx/service/OperatorKeyService.java:16]: Unnecessary Import

File: src/main/java/com/uid2/admin/Main.java (hunk 1 above)
Comment (lizk886) [reply src/main/java/com/uid2/admin/Main.java]: done

File: src/main/java/com/uid2/admin/managers/S3KeyManager.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    private final RotatingS3KeyProvider s3KeyProvider;
+    private final S3KeyStoreWriter s3KeyStoreWriter;
+    private final IKeyGenerator keyGenerator;
+
+
+    public S3KeyManager(RotatingS3KeyProvider s3KeyProvider, S3KeyStoreWriter s3KeyStoreWriter, IKeyGenerator keyGenerator) {
+        this.s3KeyProvider = s3KeyProvider;
+        this.s3KeyStoreWriter = s3KeyStoreWriter;
+        this.keyGenerator = keyGenerator;
+
```
Comment (Ian-Nara) [inline src/main/java/com/uid2/admin/managers/S3KeyManager.java:27]: empty newlines

File: src/main/java/com/uid2/admin/managers/S3KeyManager.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+    }
+
+    public S3Key generateS3Key(int siteId, long activates, long created) throws Exception {
+        int newKeyId = getNextKeyId();
+        S3Key newS3Key = new S3Key();
+        newS3Key.setId(newKeyId);
```
Comment (Ian-Nara) [inline src/main/java/com/uid2/admin/managers/S3KeyManager.java:34]: Can we pass these fields through the constructor and remove the setters? Unless there will be a reason to change the values later. Key_id and site_id at least should not be updatable after creation.

Comment (lizk886) [reply src/main/java/com/uid2/admin/managers/S3KeyManager.java:34]: right, I will keep the setter for activation and remove others

File: src/main/java/com/uid2/admin/Main.java (hunk 2 above)
Comment (lizk886) [reply src/main/java/com/uid2/admin/Main.java:206]: will do

File: src/main/java/com/uid2/admin/managers/S3KeyManager.java (hunk 4 above)
Comment (lizk886) [reply src/main/java/com/uid2/admin/managers/S3KeyManager.java:29]: you are right, no

**Review by cody-constine-ttd — APPROVED** [review]
LGTM

File: src/main/java/com/uid2/admin/Main.java [hunk 11]
```diff
@@ -26,6 +27,7 @@
 import com.uid2.admin.vertx.service.*;
 import com.uid2.shared.Const;
 import com.uid2.shared.Utils;
+import com.uid2.shared.auth.OperatorKey;
```
Comment (lizk886) [inline src/main/java/com/uid2/admin/Main.java:30]: seems like this is useless, will delete it

File: src/main/java/com/uid2/admin/Main.java [hunk 12]
```diff
@@ -61,6 +63,7 @@
 
 import javax.management.*;
 import java.lang.management.ManagementFactory;
+import java.time.Instant;
```
Comment (lizk886) [inline src/main/java/com/uid2/admin/Main.java:66]: seems like this is useless, will delete it

File: src/main/java/com/uid2/admin/Main.java [hunk 13]
```diff
@@ -195,6 +196,23 @@ public void run() {
...
             OperatorKeyStoreWriter operatorKeyStoreWriter = new OperatorKeyStoreWriter(operatorKeyProvider, fileManager, jsonWriter, versionGenerator);
 
+            CloudPath s3KeyMetadataPath = new CloudPath(config.getString(Const.Config.S3keysMetadataPathProp));
+            GlobalScope s3KeyGlobalScope = new GlobalScope(s3KeyMetadataPath);
+            RotatingS3KeyProvider s3KeyProvider = new RotatingS3KeyProvider(cloudStorage, s3KeyGlobalScope);
+            S3KeyStoreWriter s3KeyStoreWriter = new S3KeyStoreWriter(s3KeyProvider, fileManager, jsonWriter, versionGenerator, clock, s3KeyGlobalScope);
+            S3KeyManager s3KeyManager = new S3KeyManager(s3KeyProvider, s3KeyStoreWriter);
+            s3KeyManager.generateKeysForOperators(operatorKeyProvider.getAll(), config.getLong("s3_key_activates_in_seconds"), config.getInteger("s3_key_count_per_site"));
+            try {
+                s3KeyProvider.loadContent(s3KeyProvider.getMetadata());
```
Comment (lizk886) [inline src/main/java/com/uid2/admin/Main.java:206]: I want to change  s3KeyProvider.loadContent(s3KeyProvider.getMetadata()); to  s3KeyProvider.loadContent(), need to edit share though. seems like too late

**Pull request merged**